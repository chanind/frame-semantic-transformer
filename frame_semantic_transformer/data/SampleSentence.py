from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Iterable, Mapping


@dataclass
class SampleSentence:
    text: str
    trigger_loc: tuple[int, int]
    frame: str
    frame_element_locs: list[tuple[int, int, str]]

    @property
    def trigger(self) -> str:
        return self.text[self.trigger_loc[0] : self.trigger_loc[1]]

    @property
    def trigger_labeled_text(self) -> str:
        pre_span = self.text[0 : self.trigger_loc[0]]
        post_span = self.text[self.trigger_loc[1] :]
        return f"{pre_span} * {self.trigger} * {post_span}"

    @property
    def frame_elements(self) -> list[tuple[str, str]]:
        return [
            (element, self.text[loc_start:loc_end])
            for (loc_start, loc_end, element) in self.frame_element_locs
        ]


def parse_samples_from_exemplars(
    exemplars: Iterable[Mapping[str, Any]]
) -> list[SampleSentence]:
    """
    Helper to parse sample sentences out of framenet exemplars, contained in lexical units
    ex: lu = fn.lus()[0]; samples = parse_samples_from_exemplars(lu.exemplars)
    """
    sample_sentences: list[SampleSentence] = []
    for exemplar in exemplars:
        for annotation in exemplar["annotationSet"]:
            if "FE" in annotation and "Target" in annotation and "frame" in annotation:
                assert len(annotation["Target"]) == 1
                assert annotation["FE"][1] == {}
                sample_sentences.append(
                    SampleSentence(
                        text=annotation["text"],
                        trigger_loc=annotation["Target"][0],
                        frame=annotation["frame"]["name"],
                        frame_element_locs=annotation["FE"][0],
                    )
                )
    return sample_sentences
