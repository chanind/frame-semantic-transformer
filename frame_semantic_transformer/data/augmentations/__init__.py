from .chain_augmentations import chain_augmentations
from .DataAugmentation import DataAugmentation
from .LowercaseAugmentation import LowercaseAugmentation
from .RemoveContractionsAugmentation import RemoveContractionsAugmentation
from .RemoveEndPunctuationAugmentation import RemoveEndPunctuationAugmentation

__all__ = (
    "chain_augmentations",
    "DataAugmentation",
    "LowercaseAugmentation",
    "RemoveContractionsAugmentation",
    "RemoveEndPunctuationAugmentation",
)
