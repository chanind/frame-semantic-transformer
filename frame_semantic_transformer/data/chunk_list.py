from __future__ import annotations
from typing import Iterator, Sequence, TypeVar

T = TypeVar("T")


def chunk_list(lst: Sequence[T], chunk_size: int) -> Iterator[Sequence[T]]:
    for i in range(0, len(lst), chunk_size):
        yield lst[i : i + chunk_size]
