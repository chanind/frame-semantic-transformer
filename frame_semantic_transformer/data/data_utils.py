from __future__ import annotations
import re
from typing import Iterator, Sequence, TypeVar
from torch import Tensor

from transformers import T5Tokenizer

from frame_semantic_transformer.constants import PADDING_LABEL_ID

T = TypeVar("T")


def chunk_list(lst: Sequence[T], chunk_size: int) -> Iterator[Sequence[T]]:
    for i in range(0, len(lst), chunk_size):
        yield lst[i : i + chunk_size]


def standardize_punct(sent: str) -> str:
    """
    Try to standardize things like "He 's a man" -> "He's a man"
    """
    updated_sent = T5Tokenizer.clean_up_tokenization(sent)
    # remove space before punct
    updated_sent = re.sub(r"([a-zA-Z0-9])\s+(\*?[.',:?])", r"\1\2", updated_sent)
    # remove repeated *'s
    updated_sent = re.sub(r"\*+", "*", updated_sent)
    # fix spaces in contractions
    updated_sent = re.sub(r"([a-zA-Z0-9])\s+(\*?n't)", r"\1\2", updated_sent)
    # remove ``
    updated_sent = re.sub(r"\s*``\s*", " ", updated_sent)
    # replace "*n't" with "n*'t, for the tokenizer"
    updated_sent = re.sub(r"\*n't", "n*'t", updated_sent)
    # put a space between * and letter chars, since it seems to work better with the tokenizer
    updated_sent = re.sub(r"\*([a-zA-Z0-9])", r"* \1", updated_sent)

    return updated_sent.strip()


def marked_string_to_locs(
    text: str, symbol: str = "*", remove_spaces: bool = True
) -> tuple[str, list[int]]:
    """
    Take a string like "He * went to the * store" and return the indices of the tagged words,
    in this case "went" and "store", and remove the tags (in this case the *'s)
    """
    output_str = ""
    remaining_str = text
    locs: list[int] = []
    symbol_index = remaining_str.find("*")

    while symbol_index != -1:
        locs.append(symbol_index + len(output_str))
        output_str += remaining_str[:symbol_index]
        remaining_str = remaining_str[symbol_index + len(symbol) :]
        if remove_spaces:
            remaining_str = remaining_str.strip()
        symbol_index = remaining_str.find("*")
    output_str += remaining_str
    return output_str, locs


def trim_batch(
    input_ids: Tensor, attention_mask: Tensor, labels: Tensor
) -> tuple[Tensor, Tensor, Tensor]:
    """
    Helper to trim a batch of inputs / labels to strip padding down to the length of the longest item in the batch
    This helps the model run faster by avoiding needing to generate up to 512 characters of padding when the meaningful
    content is much shorter.
    """
    longest_inputs = int(attention_mask.sum(dim=1).max().item())
    longest_labels = int((labels != PADDING_LABEL_ID).sum(dim=1).max().item())

    truncated_input_ids = input_ids[:, :longest_inputs].contiguous()
    truncated_attention_mask = attention_mask[:, :longest_inputs].contiguous()
    truncated_labels = labels[:, :longest_labels].contiguous()

    return (truncated_input_ids, truncated_attention_mask, truncated_labels)
