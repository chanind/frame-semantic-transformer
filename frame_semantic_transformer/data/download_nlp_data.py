import nltk


def ensure_nlp_data_downloaded() -> None:
    for corpus in ["framenet_v17"]:
        try:
            nltk.data.find(f"corpora/{corpus}")
        except LookupError:
            nltk.download(corpus)
