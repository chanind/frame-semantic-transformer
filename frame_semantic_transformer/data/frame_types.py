from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Frame:
    """
    Representation of a FrameNet frame
    For training on your own data, you can use this class to represent your own frames
    """

    name: str
    core_elements: list[str]
    non_core_elements: list[str]
    lexical_units: list[str]


@dataclass
class FrameAnnotatedSentence:
    """
    Representation of a sentence with annotations for use in training
    If training on your own data, you'll need to create instances of this class for your training sentences
    """

    text: str
    annotations: list[FrameAnnotation]
    # if this text isn't annotated with every trigger loc, we shouldn't generate a trigger id task from it
    # but it's still useful for frame classification and argument extraction tasks
    skip_trigger_identification_task: bool = False


@dataclass
class FrameAnnotation:
    """
    A single frame occuring in a sentence
    """

    frame: str
    trigger_locs: list[int]
    frame_elements: list[FrameElementAnnotation]


@dataclass
class FrameElementAnnotation:
    """
    A single frame element in a frame annotation.
    Includes the name of the frame element and the start and end locations of the frame element in the sentence
    """

    name: str
    start_loc: int
    end_loc: int
