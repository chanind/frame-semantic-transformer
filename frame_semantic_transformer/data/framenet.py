from __future__ import annotations
from functools import lru_cache
from typing import Any, Sequence, Mapping
import nltk

from nltk.corpus import framenet as fn


class InvalidFrameError(Exception):
    pass


def ensure_framenet_downloaded() -> None:
    try:
        nltk.data.find("corpora/framenet_v17")
    except LookupError:
        nltk.download("framenet_v17")


def is_valid_frame(frame: str) -> bool:
    return frame in get_all_valid_frame_names()


def get_core_frame_elements(frame: str) -> list[str]:
    if not is_valid_frame(frame):
        raise InvalidFrameError(frame)
    return get_frame_elements_map_by_core_type()[frame]["core"]


def get_non_core_frame_elements(frame: str) -> list[str]:
    if not is_valid_frame(frame):
        raise InvalidFrameError(frame)
    return get_frame_elements_map_by_core_type()[frame]["noncore"]


@lru_cache(1)
def get_frame_elements_map_by_core_type() -> dict[str, dict[str, list[str]]]:
    """
    fast-lookup helper for frames -> frame elements for faster eval
    """
    results: dict[str, dict[str, list[str]]] = {}
    for frame in fn.frames():
        results[frame.name] = {"core": [], "noncore": []}
        for element_name, element in frame.FE.items():
            element_type = "core" if element.coreType == "Core" else "noncore"
            results[frame.name][element_type].append(element_name)
    return results


@lru_cache(1)
def get_all_valid_frame_names() -> set[str]:
    return {frame.name for frame in fn.frames()}


def get_fulltext_docs() -> Sequence[Mapping[str, Any]]:
    return fn.docs()
