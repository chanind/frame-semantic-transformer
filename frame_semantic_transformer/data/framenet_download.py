import nltk


def ensure_framenet_downloaded() -> None:
    nltk.download("framenet_v17")
