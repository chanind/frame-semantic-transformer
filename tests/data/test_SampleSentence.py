from __future__ import annotations
from syrupy.assertion import SnapshotAssertion
from nltk.corpus import framenet as fn

from frame_semantic_transformer.data.SampleSentence import (
    SampleSentence,
    parse_samples_from_fulltext_doc,
    parse_samples_from_lexical_unit,
)


def test_parse_samples_from_lexical_unit(snapshot: SnapshotAssertion) -> None:
    lu = fn.lu(6403)  # repair.v

    samples = parse_samples_from_lexical_unit(lu)

    assert len(samples) == 9
    assert (
        samples[0].text
        == "The word is that the usurper , or those acting in his name , have sent out a call for all Scots lords and landed men to repair there , to Annan , to do homage to him . "
    )
    assert (
        samples[0].trigger_labeled_text
        == "The word is that the usurper , or those acting in his name , have sent out a call for all Scots lords and landed men to * repair * there , to Annan , to do homage to him . "
    )

    assert samples[0].trigger == "repair"
    assert samples[0].frame_elements == [
        ("Self_mover", "all Scots lords and landed men"),
        ("Goal", "there , to Annan ,"),
        ("Purpose", "to do homage to him"),
    ]
    assert (
        samples[0].frame_elements_str
        == "Self_mover = all Scots lords and landed men | Goal = there , to Annan , | Purpose = to do homage to him"
    )
    assert samples == snapshot


def test_parse_samples_from_fulltext_doc(snapshot: SnapshotAssertion) -> None:
    doc = fn.docs()[17]  # random document
    samples = parse_samples_from_fulltext_doc(doc)
    assert samples == snapshot
