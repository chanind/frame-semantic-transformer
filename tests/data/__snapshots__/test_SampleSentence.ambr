# name: test_parse_samples_from_fulltext_doc
  list([
    SampleSentence(text='March 17 , 2000', trigger_loc=(0, 5), frame='Calendric_unit', frame_element_locs=[(0, 5, 'Unit')]),
    SampleSentence(text='When Jerry was released from prison , he knew it would take hard work and determination to straighten out his life .', trigger_loc=(0, 4), frame='Temporal_collocation', frame_element_locs=[(5, 35, 'Landmark_event'), (38, 114, 'Trajector_event')]),
    SampleSentence(text='When Jerry was released from prison , he knew it would take hard work and determination to straighten out his life .', trigger_loc=(29, 35), frame='Prison', frame_element_locs=[(29, 35, 'Penal_institution')]),
    SampleSentence(text='When Jerry was released from prison , he knew it would take hard work and determination to straighten out his life .', trigger_loc=(41, 45), frame='Awareness', frame_element_locs=[(38, 40, 'Cognizer'), (46, 114, 'Content')]),
    SampleSentence(text='When Jerry was released from prison , he knew it would take hard work and determination to straighten out his life .', trigger_loc=(55, 59), frame='Have_as_requirement', frame_element_locs=[(60, 87, 'Requirement'), (88, 114, 'Dependent')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(3, 7), frame='Awareness', frame_element_locs=[(0, 2, 'Cognizer'), (8, 33, 'Content')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(15, 19), frame='Intentionally_create', frame_element_locs=[(8, 10, 'Creator'), (20, 33, 'Created_entity')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(22, 25), frame='Desirability', frame_element_locs=[(26, 33, 'Evaluee')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(52, 62), frame='Commitment', frame_element_locs=[(48, 51, 'Speaker'), (63, 85, 'Message')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(104, 109), frame='Reliance_on_expectation', frame_element_locs=[(88, 90, 'Cognizer'), (110, 162, 'Expectation')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(158, 162), frame='Path_traveled', frame_element_locs=[(154, 157, 'Theme'), (158, 162, 'Path')]),
    SampleSentence(text="He knew he had made a bad mistake , but despite his commitment to rebuilding his life , he just did n't count on the obstacles that would be thrown up in his path .", trigger_loc=(151, 153), frame='Interior_profile_relation', frame_element_locs=[(113, 126, 'Figure'), (127, 150, 'Figure'), (154, 162, 'Ground')]),
    SampleSentence(text='Reality hit when he was unable to reconcile with his wife and ended up on the street .', trigger_loc=(12, 16), frame='Temporal_collocation', frame_element_locs=[(0, 11, 'Trajector_event'), (17, 84, 'Landmark_event')]),
    SampleSentence(text='Reality hit when he was unable to reconcile with his wife and ended up on the street .', trigger_loc=(24, 30), frame='Capability', frame_element_locs=[(17, 19, 'Entity'), (31, 57, 'Event')]),
    SampleSentence(text='Reality hit when he was unable to reconcile with his wife and ended up on the street .', trigger_loc=(53, 57), frame='Personal_relationship', frame_element_locs=[(49, 52, 'Partner_1'), (53, 57, 'Partner_2')]),
    SampleSentence(text='Reality hit when he was unable to reconcile with his wife and ended up on the street .', trigger_loc=(62, 70), frame='Transition_to_state', frame_element_locs=[(17, 19, 'Entity'), (71, 84, 'Final_situation')]),
    SampleSentence(text='Reality hit when he was unable to reconcile with his wife and ended up on the street .', trigger_loc=(78, 84), frame='Roadways', frame_element_locs=[(78, 84, 'Roadway')]),
    SampleSentence(text='He went to the Center Township Trustee for food stamps until he could find a job and a place to live .', trigger_loc=(3, 7), frame='Motion', frame_element_locs=[(0, 2, 'Theme'), (8, 38, 'Goal'), (39, 54, 'Purpose'), (55, 100, 'Duration')]),
    SampleSentence(text='He went to the Center Township Trustee for food stamps until he could find a job and a place to live .', trigger_loc=(55, 60), frame='Time_vector', frame_element_locs=[(0, 54, 'Event'), (61, 100, 'Landmark_event')]),
    SampleSentence(text='He went to the Center Township Trustee for food stamps until he could find a job and a place to live .', trigger_loc=(70, 74), frame='Locating', frame_element_locs=[(61, 63, 'Perceiver'), (75, 100, 'Sought_entity')]),
    SampleSentence(text='He went to the Center Township Trustee for food stamps until he could find a job and a place to live .', trigger_loc=(87, 92), frame='Locale', frame_element_locs=[(87, 92, 'Locale')]),
    SampleSentence(text="The Trustee referred him to the Blue Triangle where he consulted Goodwill 's Career Academy staff for job search assistance .", trigger_loc=(46, 51), frame='Locative_relation', frame_element_locs=[(28, 45, 'Ground'), (52, 123, 'Figure')]),
    SampleSentence(text="The Trustee referred him to the Blue Triangle where he consulted Goodwill 's Career Academy staff for job search assistance .", trigger_loc=(92, 97), frame='Employing', frame_element_locs=[(65, 91, 'Employer'), (92, 97, 'Employee')]),
    SampleSentence(text="But Jerry 's troubles did n't end there .", trigger_loc=(13, 21), frame='Predicament', frame_element_locs=[(4, 12, 'Experiencer'), (13, 21, 'Situation')]),
    SampleSentence(text="But Jerry 's troubles did n't end there .", trigger_loc=(30, 33), frame='Process_end', frame_element_locs=[(4, 21, 'Process'), (34, 39, 'Place')]),
    SampleSentence(text='The seasonal construction job he obtained soon ended , and he returned to Career Academy for help once more .', trigger_loc=(33, 41), frame='Getting', frame_element_locs=[(0, 29, 'Theme'), (30, 32, 'Recipient')]),
    SampleSentence(text='The seasonal construction job he obtained soon ended , and he returned to Career Academy for help once more .', trigger_loc=(47, 52), frame='Process_end', frame_element_locs=[(0, 41, 'Process'), (42, 46, 'Time')]),
    SampleSentence(text='The seasonal construction job he obtained soon ended , and he returned to Career Academy for help once more .', trigger_loc=(62, 70), frame='Arriving', frame_element_locs=[(59, 61, 'Theme'), (71, 88, 'Goal'), (89, 97, 'Purpose'), (98, 107, 'Frequency')]),
    SampleSentence(text='Unemployed , homeless and with a prison record , Jerry now faced going back to prison for failure to pay child support .', trigger_loc=(33, 39), frame='Prison', frame_element_locs=[(33, 39, 'Penal_institution')]),
    SampleSentence(text='Unemployed , homeless and with a prison record , Jerry now faced going back to prison for failure to pay child support .', trigger_loc=(55, 58), frame='Temporal_collocation', frame_element_locs=[(49, 54, 'Trajector_event'), (59, 85, 'Trajector_event')]),
    SampleSentence(text='Unemployed , homeless and with a prison record , Jerry now faced going back to prison for failure to pay child support .', trigger_loc=(65, 70), frame='Motion', frame_element_locs=[(49, 54, 'Theme'), (55, 58, 'Time'), (71, 75, 'Direction'), (76, 85, 'Goal')]),
    SampleSentence(text='Unemployed , homeless and with a prison record , Jerry now faced going back to prison for failure to pay child support .', trigger_loc=(79, 85), frame='Prison', frame_element_locs=[(79, 85, 'Penal_institution')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(8, 12), frame='Proportional_quantity', frame_element_locs=[(8, 12, 'Denoted_quantity'), (13, 17, 'Mass')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(76, 80), frame='Capability', frame_element_locs=[(66, 71, 'Entity'), (81, 191, 'Event')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(103, 107), frame='Project', frame_element_locs=[(66, 71, 'Agent'), (95, 102, 'Goal')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(140, 144), frame='Locating', frame_element_locs=[(66, 71, 'Perceiver'), (145, 152, 'Sought_entity')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(145, 152), frame='Buildings', frame_element_locs=[(145, 152, 'Building')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(157, 164), frame='Intentionally_act', frame_element_locs=[(66, 71, 'Agent'), (165, 191, 'Act')]),
    SampleSentence(text="It took some time and hard work , but with the help of Goodwill , Jerry was able to work out a payment plan with the prosecutor 's office , find housing and conduct a more thorough job search .", trigger_loc=(3, 7), frame='Have_as_requirement', frame_element_locs=[(0, 2, 'Dependent'), (8, 31, 'Requirement')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(25, 31), frame='Position_on_a_scale', frame_element_locs=[(25, 31, 'Value'), (32, 37, 'Item')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(41, 46), frame='Measure_duration', frame_element_locs=[(41, 46, 'Unit')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(47, 49), frame='Interior_profile_relation', frame_element_locs=[(5, 46, 'Figure'), (50, 65, 'Ground')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(68, 72), frame='Quantified_mass', frame_element_locs=[(68, 72, 'Quantity'), (73, 79, 'Individuals')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(73, 79), frame='People', frame_element_locs=[(73, 79, 'Person')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(80, 85), frame='Continued_state_of_affairs', frame_element_locs=[(68, 79, 'State_of_affairs'), (80, 85, 'Reference_occasion'), (86, 132, 'State_of_affairs')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(86, 90), frame='Possession', frame_element_locs=[(68, 79, 'Owner'), (80, 85, 'Time'), (91, 132, 'Possession')]),
    SampleSentence(text='With unemployment at its lowest point in years in central Indiana , many people still have a hard time finding meaningful employment .', trigger_loc=(93, 97), frame='Difficulty', frame_element_locs=[(68, 79, 'Experiencer'), (98, 132, 'Activity')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(7, 11), frame='Increment', frame_element_locs=[(12, 54, 'Class')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(24, 30), frame='People', frame_element_locs=[(24, 30, 'Person'), (31, 54, 'Descriptor')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(40, 46), frame='Assistance', frame_element_locs=[(3, 30, 'Benefited_party'), (31, 39, 'Helper'), (47, 54, 'Time')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(47, 49), frame='Temporal_collocation', frame_element_locs=[(3, 46, 'Trajector_event'), (50, 54, 'Landmark_period')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(57, 61), frame='Proportional_quantity', frame_element_locs=[(57, 61, 'Denoted_quantity'), (62, 75, 'Individuals')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(76, 80), frame='Possession', frame_element_locs=[(57, 75, 'Owner'), (81, 127, 'Possession')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(81, 89), frame='Quantified_mass', frame_element_locs=[(81, 89, 'Quantity'), (90, 127, 'Individuals')]),
    SampleSentence(text='Of the more than 42,000 people Goodwill served in 1999 , over 5,000 of them have multiple barriers to finding and keeping a job .', trigger_loc=(68, 70), frame='Partitive', frame_element_locs=[(57, 67, 'Subset'), (71, 75, 'Group')]),
    SampleSentence(text='Some are blind , deaf or have other physical disabilities .', trigger_loc=(30, 35), frame='Increment', frame_element_locs=[(36, 57, 'Class')]),
    SampleSentence(text='More than 100 were homeless .', trigger_loc=(0, 4), frame='Increment', frame_element_locs=[(5, 13, 'Class')]),
    SampleSentence(text='But Goodwill offers hope even to people with numerous obstacles to overcome , like Jerry .', trigger_loc=(33, 39), frame='People', frame_element_locs=[(33, 39, 'Person'), (40, 75, 'Descriptor')]),
    SampleSentence(text='But Goodwill offers hope even to people with numerous obstacles to overcome , like Jerry .', trigger_loc=(45, 53), frame='Quantified_mass', frame_element_locs=[(45, 53, 'Quantity'), (54, 75, 'Individuals')]),
    SampleSentence(text='But Goodwill offers hope even to people with numerous obstacles to overcome , like Jerry .', trigger_loc=(78, 82), frame='Similarity', frame_element_locs=[(33, 75, 'Entity_1'), (83, 88, 'Entity_2')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(17, 20), frame='Age', frame_element_locs=[(17, 20, 'Age'), (21, 68, 'Entity')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(25, 27), frame='Locative_relation', frame_element_locs=[(5, 24, 'Figure'), (28, 68, 'Ground')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(56, 68), frame='Manufacturing', frame_element_locs=[(30, 55, 'Product'), (56, 68, 'Producer')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(81, 89), frame='Predicament', frame_element_locs=[(71, 74, 'Experiencer'), (75, 80, 'Domain')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(110, 115), frame='Locale', frame_element_locs=[(110, 115, 'Locale')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(132, 135), frame='Capability', frame_element_locs=[(126, 131, 'Entity'), (136, 225, 'Event')]),
    SampleSentence(text='With a promising new job at a plastic injection molding manufacturer , his legal troubles in check and a safe place to live , Jerry can get on with the business of building a life for himself and contributing to his community .', trigger_loc=(216, 225), frame='Aggregate', frame_element_locs=[(216, 225, 'Aggregate')]),
    SampleSentence(text='For as long as he needs it , Goodwill will continue to help .', trigger_loc=(43, 51), frame='Activity_ongoing', frame_element_locs=[(29, 37, 'Agent'), (52, 59, 'Activity')]),
    SampleSentence(text='For as long as he needs it , Goodwill will continue to help .', trigger_loc=(18, 23), frame='Needing', frame_element_locs=[(15, 17, 'Cognizer'), (24, 26, 'Requirement')]),
    SampleSentence(text='Your support of Goodwill will provide job training and placement services to help the hardest to serve in central Indiana find meaningful employment .', trigger_loc=(65, 73), frame='Public_services', frame_element_locs=[]),
    SampleSentence(text='Your support of Goodwill will provide job training and placement services to help the hardest to serve in central Indiana find meaningful employment .', trigger_loc=(97, 102), frame='Assistance', frame_element_locs=[]),
    SampleSentence(text='Your support of Goodwill will provide job training and placement services to help the hardest to serve in central Indiana find meaningful employment .', trigger_loc=(122, 126), frame='Locating', frame_element_locs=[]),
    SampleSentence(text='Your support of Goodwill will provide job training and placement services to help the hardest to serve in central Indiana find meaningful employment .', trigger_loc=(138, 148), frame='Being_employed', frame_element_locs=[]),
  ])
# ---
# name: test_parse_samples_from_lexical_unit
  list([
    SampleSentence(text='The word is that the usurper , or those acting in his name , have sent out a call for all Scots lords and landed men to repair there , to Annan , to do homage to him . ', trigger_loc=(120, 126), frame='Self_motion', frame_element_locs=[(86, 116, 'Self_mover'), (127, 145, 'Goal'), (146, 165, 'Purpose')]),
    SampleSentence(text='The latter were forced to repair to the waiting area where they enviously sipped ginger beer while waiting their turn and watching the old horses pull the machines into the water , one after another . ', trigger_loc=(26, 32), frame='Self_motion', frame_element_locs=[(0, 10, 'Self_mover'), (33, 198, 'Goal')]),
    SampleSentence(text='To do the Cafe Posta properly , one repairs to the antique-filled back room ; front parlour strictly for tourists . ', trigger_loc=(36, 43), frame='Self_motion', frame_element_locs=[(0, 29, 'Purpose'), (32, 35, 'Self_mover'), (44, 75, 'Goal')]),
    SampleSentence(text='Two of the cast fainted and most of the rest repaired to the nearest bar . ', trigger_loc=(45, 53), frame='Self_motion', frame_element_locs=[(28, 44, 'Self_mover'), (54, 72, 'Goal')]),
    SampleSentence(text="Shortly they 'll both repair to the bedroom for another three minute interview break . ", trigger_loc=(22, 28), frame='Self_motion', frame_element_locs=[(0, 7, 'Time'), (8, 21, 'Self_mover'), (29, 43, 'Goal'), (44, 84, 'Purpose')]),
    SampleSentence(text='The service over , Joshua repaired to the clinically clean Gents . ', trigger_loc=(26, 34), frame='Self_motion', frame_element_locs=[(0, 16, 'Time'), (19, 25, 'Self_mover'), (35, 64, 'Goal')]),
    SampleSentence(text='He repaired to his kitchen area , flagrantly disregarding the woman who was being sick three rows behind the two men . ', trigger_loc=(3, 11), frame='Self_motion', frame_element_locs=[(0, 2, 'Self_mover'), (12, 31, 'Goal')]),
    SampleSentence(text='He and Sergeant Joe , having met in Petticoat Lane and having talked , repaired to a pub and had a long and very friendly conversation . ', trigger_loc=(71, 79), frame='Self_motion', frame_element_locs=[(0, 19, 'Self_mover'), (22, 68, 'Time'), (80, 88, 'Goal')]),
    SampleSentence(text='Fishermen would repair to the woods to cut hazel and withies to make or mend their pots , then set out in their small boats to their chosen ` grounds " to drop the pots . ', trigger_loc=(16, 22), frame='Self_motion', frame_element_locs=[(0, 9, 'Self_mover'), (23, 35, 'Goal'), (36, 87, 'Purpose')]),
  ])
# ---
